File: utils/psnr.py
import cv2
import numpy as np
import math


def calculate_psnr(img1_path: str, img2_path: str) -> float:
    # Đọc ảnh (định dạng mặc định BGR của OpenCV)
    img1 = cv2.imread(img1_path)
    img2 = cv2.imread(img2_path)

    if img1 is None or img2 is None:
        raise FileNotFoundError("Không thể đọc được ảnh từ đường dẫn cung cấp.")

    # Kiểm tra kích thước ma trận ảnh
    if img1.shape != img2.shape:
        raise ValueError("Hai ảnh phải có cùng kích thước (width, height, channels).")

    # Ép kiểu dữ liệu sang float64 để tránh tràn bộ nhớ (overflow) khi tính bình phương
    img1_float = img1.astype(np.float64)
    img2_float = img2.astype(np.float64)

    # Tính Mean Squared Error (MSE) cho toàn bộ ma trận
    mse = np.mean((img1_float - img2_float) ** 2)

    # Nếu MSE = 0, hai ảnh giống hệt nhau, PSNR tiến tới vô cực
    if mse == 0:
        return float('inf')

    # Giá trị pixel tối đa của ảnh 8-bit
    max_pixel = 255.0

    # Áp dụng công thức tính PSNR
    psnr = 10 * math.log10((max_pixel ** 2) / mse)
    return psnr

# Cách gọi hàm
# psnr_value = calculate_psnr("original.png", "processed.png")
# print(f"PSNR: {psnr_value:.2f} dB")

File: csrc/psnr.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <cmath>

using namespace cv;
using namespace std;

double calculatePSNR(const string& img1_path, const string& img2_path) {
    // Đọc ảnh
    Mat img1 = imread(img1_path);
    Mat img2 = imread(img2_path);

    if (img1.empty() || img2.empty()) {
        cerr << "Lỗi: Không thể tải ảnh." << endl;
        return -1.0;
    }

    if (img1.rows != img2.rows || img1.cols != img2.cols || img1.type() != img2.type()) {
        cerr << "Lỗi: Kích thước hoặc định dạng của hai ảnh không khớp." << endl;
        return -1.0;
    }

    Mat s1;
    // Lấy giá trị tuyệt đối của hiệu hai ma trận: |img1 - img2|
    absdiff(img1, img2, s1);

    // Ép kiểu sang Float 32-bit để bình phương không bị tràn dải 8-bit
    s1.convertTo(s1, CV_32F);

    // Bình phương từng phần tử: |img1 - img2|^2
    s1 = s1.mul(s1);

    // Tính tổng các sai số bình phương theo từng kênh màu
    Scalar s = sum(s1);
    double sse = s.val[0] + s.val[1] + s.val[2];

    // Nếu Sum of Squared Errors tiến tới 0 (ảnh giống nhau hoàn toàn)
    if (sse <= 1e-10) {
        return INFINITY;
    } else {
        // Tính MSE = Tổng sai số / (Tổng số pixel * Số kênh màu)
        double mse = sse / (double)(img1.channels() * img1.total());

        // Tính PSNR với MAX_I = 255
        double psnr = 10.0 * log10((255 * 255) / mse);
        return psnr;
    }
}

int main() {
    /* string path_original = "original.png";
    string path_processed = "processed.png";

    double psnr_val = calculatePSNR(path_original, path_processed);
    if (psnr_val != -1.0) {
        cout << "PSNR: " << psnr_val << " dB" << endl;
    }
    */
    return 0;
}